### PR TITLE
Fix error reporting of templates

### PR DESCRIPTION
**Pull request by MrAnno**
`log_template_compile()` has a return value, which can indicate compilation error.

Unfortunately, the return value was not checked in several places, producing repeated run-time errors and unexpected behavior.

For example:
```
options {
	create-dirs(yes);
};

log {
	source { stdin(); };
	destination {
		file("/tmp/${DATE");
	};

	flags(flow-control);
};
```

The above config snippet created a file named `${DATE` inside the following directory structure:
```
$ tree .

error in template: /
└── tmp
    └── ${DATE
```

This is because the `log_template_format()` method called on an invalid template will contain the error itself (`error in template: ...`).

This PR adds proper error handling where templates are used, or where possible, just replaces the template creation logic with the `template_content` grammar rule, which takes care of error reporting:

```
Error parsing affile, Error compiling template, error=Invalid macro, '}' is missing, error_pos='11' in /home/anno/dev/syslog-ng/install/etc/syslog-ng.conf:39:8-39:21:
34      };
35
36      log {
37              source { stdin(); };
38              destination {
39---->                 file("/tmp/${DATE");
39---->                      ^^^^^^^^^^^^^
```

The PR contains some additional unit tests for non-trivial templates, and also makes empty templates (`template("")`) trivial and adds `log_template_is_literal_string()`, so we can get rid of the `strchr($)` trickery in the file destination.

### CODE REVIEW

**Comment by kira-syslogng** [comment]
Build SUCCESS

File: modules/affile/affile-dest.c [hunk 1]
```diff
@@ -769,14 +769,13 @@ affile_dd_new_instance(gchar *filename, GlobalConfig *cfg)
...
-  self->filename_template = log_template_new(cfg, NULL);
-  log_template_compile(self->filename_template, filename, NULL);
+  self->filename_template = filename_template;
   log_writer_options_defaults(&self->writer_options);
   self->writer_options.mark_mode = MM_NONE;
   self->writer_options.stats_level = STATS_LEVEL1;
   self->writer_flags = LW_FORMAT_FILE;
 
-  if (strchr(filename, '$') != NULL)
+  if (strchr(filename_template->template, '$') != NULL)
```
Comment (Kokan) [inline modules/affile/affile-dest.c:778]: Hmm, this little lie starts to get dirty. I mean, the filename in fact always a template. Rather that template became constant.

Comment (MrAnno) [inline modules/affile/affile-dest.c:778]: This looks pretty bad (@Kokan noticed that too). After my change, the layering violation became clear.

I was planning to add a `log_template_is_literal_string()`/`log_template_get_literal()` pair to the template API for #3621.

I'll push the commit containing those methods into this branch instead, so we can replace the `strchr()` call with `log_template_is_literal_string()`, which makes sure the template is really a literal string in the appropriate layer (without using `strchr()`).

Comment (MrAnno) [reply modules/affile/affile-dest.c:778]: Yeah, this looks pretty bad. After my change, the layering violation became clear.

I was planning to add a log_template_is_literal_string()/log_template_get_literal() pair to the template API for #3621.

I'll push the commit containing those methods into this branch instead, so we can replace the strchr() call with log_template_is_literal_string(), which makes sure the template is really a literal string in the appropriate layer (without using strchr()).

File: lib/template/tests/test_template.c [hunk 2]
```diff
@@ -354,6 +354,16 @@ Test(template, test_single_values_and_literal_strings_are_considered_trivial)
   LogTemplate *template;
   LogMessage *msg = create_sample_message();
 
+  template = compile_template("", FALSE);
```
[the remark this answers is not in the document]
Comment (MrAnno) [reply lib/template/tests/test_template.c:357]: Good idea, thanks. I'll add that check.

File: modules/affile/affile-dest.c (hunk 1 above)
Comment (MrAnno) [reply modules/affile/affile-dest.c]: Done.

**Comment by MrAnno** [comment]
@kira-syslogng do stresstest

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by kira-syslogng** [comment]
Kira-stress-test: Build SUCCESS

**Review by Kokan — APPROVED** [review]
Thanks.

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Review by Kokan — DISMISSED** [review]

**Review by gaborznagy — COMMENTED** [review]
Besides a minor thing, approve. :+1:

File: lib/template/tests/test_template.c [hunk 3]
```diff
@@ -349,35 +349,122 @@ Test(template, test_template_function_args)
...
+  assert_template_trivial_value(" ", msg, " ");
+  assert_template_trivial_value("literal", msg, "literal");
+  assert_template_trivial_value("$1", msg, "first-match");
+  assert_template_trivial_value("$MSG", msg, "árvíztűrőtükörfúrógép");
+  assert_template_trivial_value("$HOST", msg, "bzorp");
+  assert_template_trivial_value("${APP.VALUE}", msg, "value");
+
+  LogTemplate *template = log_template_new(configuration, NULL);
+  cr_assert_not(log_template_compile(template, "$1 $2 ${MSG invalid", NULL));
+  cr_assert_(log_template_is_trivial(template), "Invalid templates are trivial");
```
[the remark this answers is not in the document]
Comment (MrAnno) [reply lib/template/tests/test_template.c:386]: Oh, I didn't even know this works, that was a typo. Thanks.

**Review by gaborznagy — APPROVED** [review]

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Review by Kokan — APPROVED** [review]

**Pull request merged**